vllm/model_executor/models/minicpmv.py

```diff
@@ -559,119 +510,72 @@ def get_image_bounds(self, input_ids: torch.Tensor):
-                        for i in range(B):
-                            patch_attn_mask[i, :tgt_sizes[i][0] *
-                                            tgt_sizes[i][1]] = True
-                        vision_embedding = self.vpm(
-                            all_pixel_values.type(dtype),
-                            patch_attention_mask=patch_attn_mask
-                        ).last_hidden_state
-                    else:
-                        for i in range(B):
-                            patch_attn_mask[i, 0, :tgt_sizes[i][0] *
-                                            tgt_sizes[i][1]] = True
-                        vision_embedding = self.vpm(
-                            all_pixel_values.type(dtype),
-                            patch_attention_mask=patch_attn_mask,
-                            tgt_sizes=tgt_sizes).last_hidden_state
-
-                    vision_hidden_states = self.resampler(
-                        vision_embedding, tgt_sizes)
-
-                else:  # no image
-                    dummy_feature = []
-                    vision_hidden_states = dummy_feature
+    def get_embedding(
+        self,
+        input_ids: torch.Tensor,
+        image_inputs: Optional[MiniCPMVImageInputs],
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if image_inputs is None:  # No image
+            vlm_embedding = torch.tensor([], device=input_ids.device)
+            vision_hidden_states = torch.tensor([], device=input_ids.device)
         else:
-            vision_hidden_states = data["vision_hidden_states"]
-
-        return vision_hidden_states
-
-    def get_embedding(self, data: Dict[str, Union[List[torch.Tensor],
-                                                  torch.Tensor]]):
-        input_ids = data["input_ids"]
-
-        vision_hidden_states = self.get_vision_hidden_states(data)
-        if vision_hidden_states is not None and len(vision_hidden_states) > 0:
+            vision_hidden_states = self.get_vision_hidden_states(image_inputs)
             image_bounds = self.get_image_bounds(input_ids)
-        else:
-            image_bounds = []
 
-        if hasattr(self.config, 'scale_emb'):
-            vlm_embedding = self.llm.embed_tokens(
-                input_ids) * self.config.scale_emb
-        else:
-            vlm_embedding = self.llm.embed_tokens(input_ids)
-        vision_hidden_states = [
-            i.type(vlm_embedding.dtype) if isinstance(i, torch.Tensor) else i
-            for i in vision_hidden_states
-        ]
+            if hasattr(self.config, "scale_emb"):
+                vlm_embedding: torch.Tensor = (
+                    self.llm.embed_tokens(input_ids) * self.config.scale_emb)
+            else:
+                vlm_embedding = self.llm.embed_tokens(input_ids)
 
-        if len(vision_hidden_states) > 0 and len(image_bounds) > 0:
-            vision_hidden_states = torch.cat(vision_hidden_states, dim=0)
             image_indices = torch.stack([
```

Comment:
Because `image_bounds` may be `[]`, here should add `if len(image_bounds) > 0`

Reply:
Can `image_bounds` be empty if `len(image_inputs["pixel_values"]) > 0`?

Reply:
In the preloading calculations, use dummy images and dummy `input_ids`. Here gets empty `image_bounds`.

Reply:
I think we should actually calculate `image_bounds` during the parsing stage to make things easier.